bluebird-promisell.js

```diff
@@ -1,56 +1,26 @@
 'use strict';
 
-const fs = require('fs');
-const path = require('path');
-
-const Promise = require('bluebird');
-const P = require('bluebird-promisell');
-const R = require('ramda');
-const S = require('sanctuary');
-
-//    data Text = Buffer | String
-//    readFile :: String -> String -> Promise Error Text
-const readFile = R.curry((encoding, filename) =>
-  new Promise((resolve, reject) => {
-    fs.readFile(filename, {encoding: encoding}, (err, data) => {
-      if (err != null) {
-        reject(err);
-      } else {
-        resolve(data);
-      }
-    });
-  })
-);
-
-//    readFileWithDirAndName :: String -> String -> String -> Promise Error String
-const readFileWithDirAndName = R.curry((encoding, dir, name) => {
-  const filePath = path.join(dir, name);
-  return readFile(encoding)(filePath);
-});
-
-//    readFilesWithDirAndNames :: String -> [String] -> Promise Error [String]
-const readFilesWithDirAndNames = R.compose(P.traversep, readFileWithDirAndName);
-
-//    concatFiles :: String -> Promise Error String
-const concatFiles = dir => {
-  const indexFileP = readFileWithDirAndName('utf8', dir, 'index.txt');
-  const fileNamesP = P.liftp1(S.lines)(indexFileP);
-  const readFilesWithNames = readFilesWithDirAndNames('utf8', dir);
-  const filesP = P.liftp1(readFilesWithNames)(fileNamesP);
-  return P.liftp1(R.join(''))(filesP);
-};
+/* eslint func-call-spacing: "off", no-unexpected-multiline: "off" */
+
+const P             = require('bluebird-promisell');
+
+const exit0         = require('./common/exit0');
+const exit1         = require('./common/exit1');
+const join          = require('./common/join');
+const readFile      = require('./common/read-file-promise');
+const S             = require('./common/sanctuary');
+
+
+//    concatFiles :: (String -> String) -> Promise Error String
+const concatFiles = path =>
+  P.liftp1(S.joinWith(''))
+          (P.liftp1(P.traversep(S.compose(readFile, path)))
+                   (P.liftp1(S.lines)
+                            (readFile(path('index.txt')))));
 
 
 const main = () => {
-  concatFiles(process.argv[2])
-  .then(data => {
-          process.stdout.write(data);
-          process.exit(0);
-        },
-        err => {
-          process.stderr.write(String(err) + '\n');
-          process.exit(1);
-        });
+  concatFiles(join(process.argv[2])).then(exit0, exit1);
 };
 
 if (process.mainModule.filename === __filename) main();
```

Comment:
Nice refactor! 

For the `concatFiles` function, it's very clean to chain all the sub functions together. However I'm afraid it would be a bit hard for learners to read.

I'd like to add some intermediate variables and type signatures so that it's clear what each step returns. 
And this style also highlights that the logic reads as if it's imperative and sync, but is actually functional and async. 

```js
//  concatFiles :: (String -> String) -> Promise Error String
const concatFiles = path => {
  // String -> Promise Error String
  const readFileFromPath = S.compose(readFile, path);
  // Promise Error String
  const indexFileP = readFileFromPath('index.txt');
  // Promise Error [String]
  const fileNamesP = P.liftp1(S.lines)(indexFileP);
  // Promise Error [String]
  const filesP = P.liftp1(P.traversep(readFileFromPath)(fileNamesP);
  // Promise Error String
  return P.liftp1(S.joinWith(''))(filesP);
}
```